diff --git a/ChangeLog b/ChangeLog
index f386e9a579..252f302a59 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,118 +1,3 @@
-2023-08-14  Bruno Haible  <bruno@clisp.org>
-
-	readutmp, boot-time: Don't use __UT_* symbols (regression 2023-08-11).
-	* lib/readutmp.h (_GL_UT_USER_SIZE, _GL_UT_ID_SIZE, _GL_UT_LINE_SIZE,
-	_GL_UT_HOST_SIZE): New macros.
-	(struct utmpx32): Use them.
-
-2023-08-14  Bruno Haible  <bruno@clisp.org>
-
-	readutmp, boot-time: Fix build on 32-bit glibc (regression 2023-08-11).
-	Reported by Andreas Schwab <schwab@suse.de> in
-	<https://lists.gnu.org/archive/html/bug-gnulib/2023-08/msg00125.html>.
-	* lib/readutmp.h (struct utmpx32): Reference __UT_NAMESIZE, not
-	__UT_USERSIZE.
-
-2023-08-13  Bruno Haible  <bruno@clisp.org>
-
-	stdint: Fix configure test result with gcc 4.7 or 4.8.
-	* m4/stdint.m4 (gl_STDINT_H): Don't assume that _Generic works with
-	GCC versions < 4.9 with -std=gnu11.
-
-2023-08-13  Bruno Haible  <bruno@clisp.org>
-
-	Fix some test module descriptions.
-	* modules/boot-time-tests (Files): Add tests/macros.h.
-	* modules/dprintf-gnu-tests (Files): Likewise.
-	* modules/fclose-tests (Files): Likewise.
-	* modules/getcwd-tests (Files): Likewise.
-	* modules/lock-tests (Files): Likewise.
-	* modules/nullptr-c++-tests (Files): Likewise.
-	* modules/posix_spawn-tests (Files): Likewise.
-	* modules/posix_spawnp-tests (Files): Likewise.
-	* modules/readutmp-tests (Files): Likewise.
-	* modules/sh-quote-tests (Files): Likewise.
-	* modules/unigbrk/u16-grapheme-breaks-tests (Files): Likewise.
-	* modules/unigbrk/u16-grapheme-next-tests (Files): Likewise.
-	* modules/unigbrk/u16-grapheme-prev-tests (Files): Likewise.
-	* modules/unigbrk/u32-grapheme-breaks-tests (Files): Likewise.
-	* modules/unigbrk/u32-grapheme-next-tests (Files): Likewise.
-	* modules/unigbrk/u32-grapheme-prev-tests (Files): Likewise.
-	* modules/unigbrk/u8-grapheme-breaks-tests (Files): Likewise.
-	* modules/unigbrk/u8-grapheme-next-tests (Files): Likewise.
-	* modules/unigbrk/u8-grapheme-prev-tests (Files): Likewise.
-	* modules/unigbrk/uc-grapheme-breaks-tests (Files): Likewise.
-
-2023-08-13  Bruno Haible  <bruno@clisp.org>
-
-	nproc, physmem: Use sysctl() as a fallback on GNU/kFreeBSD.
-	* lib/nproc.c: Do include <sys/sysctl.h> on GNU/kFreeBSD.
-	(num_processors_ignoring_omp): Call sysctl on GNU/kFreeBSD.
-	* lib/physmem.c: Do include <sys/sysctl.h> on GNU/kFreeBSD.
-	(physmem_total, physmem_available): Call sysctl on GNU/kFreeBSD.
-
-2023-08-13  Bruno Haible  <bruno@clisp.org>
-
-	physmem: Add tests.
-	* tests/test-physmem.c: New file.
-	* modules/physmem-tests: New file.
-
-2023-08-13  Bruno Haible  <bruno@clisp.org>
-
-	readutmp, boot-time: Fix warning on glibc 2.30..2.31 on Linux.
-	Reported by Pádraig Brady in
-	<https://lists.gnu.org/archive/html/bug-gnulib/2023-08/msg00110.html>.
-	* lib/readutmp.c: Don't include <sys/sysctl.h> on glibc/Linux.
-	* lib/boot-time.c: Likewise.
-
-2023-08-13  Paul Eggert  <eggert@cs.ucla.edu>
-
-	fts: improve memory-allocation errno handling
-	* lib/fts.c (fts_read, fts_build): When enter_dir or setup_dir
-	fails, report its errno value back to the user.
-	(internal_function): Preserve errno.
-	* modules/fts (Depends-on): Add calloc-posix, free-posix,
-	malloc-posix, and realloc-posix, for better errno handling
-	on allocation failures.
-
-	hash: set errno on failure
-	* lib/hash.c: Include errno.h.
-	(compute_bucket_size, hash_initialize, hash_rehash)
-	(hash_insert_if_absent): Set errno reliably on failure.
-	(hash_free): Preserve errno, like plain 'free'.
-	* modules/hash (Depends-on): Depend on calloc-posix,
-	free-posix, malloc-posix, so that errno is set reliably.
-
-2023-08-13  Bruno Haible  <bruno@clisp.org>
-
-	readutmp, boot-time: Fix compilation error on old Android.
-	Reported by Po Lu in
-	<https://lists.gnu.org/archive/html/bug-gnulib/2023-08/msg00108.html>.
-	* lib/readutmp.h (BOOT_TIME): Add fallback.
-
-2023-08-13  Bruno Haible  <bruno@clisp.org>
-
-	readutmp, boot-time: Fix compilation error on Android API 8.
-	Reported by Po Lu in
-	<https://lists.gnu.org/archive/html/bug-gnulib/2023-08/msg00108.html>.
-	* m4/readutmp.m4 (gl_PREREQ_READUTMP_H): Test whether sysinfo is
-	declared.
-	* lib/boot-time-aux.h (get_linux_uptime): Invoke sysinfo only if it is
-	declared.
-	* doc/glibc-functions/sysinfo.texi: Mention the Android problem.
-
-2023-08-13  Bruno Haible  <bruno@clisp.org>
-
-	readutmp, boot-time: Fix compilation error on old Android.
-	Reported by Po Lu in
-	<https://lists.gnu.org/archive/html/bug-gnulib/2023-08/msg00108.html>.
-	* lib/readutmp.c (endutent): New fallback declaration, for Android.
-	(getutent): Remove Ultrix workaround from 2000-04-05.
-	* lib/boot-time.c: Likewise.
-	* m4/readutmp.m4 (gl_PREREQ_READUTMP_H): Test whether endutent is
-	declared, not getutent.
-	* doc/glibc-functions/endutent.texi: Mention the Android bug.
-
 2023-08-12  Paul Eggert  <eggert@cs.ucla.edu>
 
 	boot-time,readutmp: do not depend on c-strtod
diff --git a/build-aux/ar-lib b/build-aux/ar-lib
index 17de55e79c..33f0b676ef 100755
--- a/build-aux/ar-lib
+++ b/build-aux/ar-lib
@@ -1,6 +1,11 @@
 #! /bin/sh
 # Wrapper for Microsoft lib.exe
+set -eo pipefail
 
+# Wrapper for compilers which do not understand '-c -o'.
+if [[ $GNU_BUILD_WRAPPER_DEBUG -eq 1 || -v COLOR_MAJOR ]]; then #if COLOR_MAJOR is set we want to include the wrapper so we can unset it otherwise would screw cmd output
+	. "$(dirname "$(readlink -f "$0")")/wrapper_helper.sh" "$@"
+fi
 me=ar-lib
 scriptversion=2019-07-04.01; # UTC
 
@@ -29,6 +34,14 @@ scriptversion=2019-07-04.01; # UTC
 # bugs to <bug-automake@gnu.org> or send patches to
 # <automake-patches@gnu.org>.
 
+func_ar_run ()
+{
+  if [[ $GNU_BUILD_WRAPPER_DEBUG -eq 1 ]]; then
+  	wrapper_exec $AR -NOLOGO "$@"
+  else
+		 exec $AR -NOLOGO "$@"
+  fi
+}
 
 # func_error message
 func_error ()
@@ -49,6 +62,10 @@ func_file_conv ()
     / | /[!/]*) # absolute file, and not a UNC file
       if test -z "$file_conv"; then
 	# lazily determine how to convert abs files
+        OS_SYS=`uname -o`
+        if [[ "$OS_SYS" == "Msys" ]]; then
+            file_conv=cygwin
+        else
 	case `uname -s` in
 	  MINGW*)
 	    file_conv=mingw
@@ -60,6 +77,7 @@ func_file_conv ()
 	    file_conv=wine
 	    ;;
 	esac
+        fi
       fi
       case $file_conv in
 	mingw)
@@ -92,7 +110,7 @@ func_at_file ()
 
   for member
   do
-    $AR -NOLOGO $operation:"$member" "$archive" || exit $?
+    func_ar_run $operation:"$member" "$archive"
   done
 }
 
@@ -205,7 +223,7 @@ if test -n "$delete"; then
         ;;
       *)
         func_file_conv "$1"
-        $AR -NOLOGO -REMOVE:"$file" "$archive" || exit $?
+        func_ar_run -REMOVE:"$file" "$archive"
         ;;
     esac
   done
@@ -223,15 +241,15 @@ elif test -n "$extract"; then
           ;;
         *)
           func_file_conv "$1"
-          $AR -NOLOGO -EXTRACT:"$file" "$archive" || exit $?
+          func_ar_run -EXTRACT:${COLOR_MAJOR}"$file"${COLOR_NONE} "$archive"
           ;;
       esac
     done
   else
-    $AR -NOLOGO -LIST "$archive" | tr -d '\r' | sed -e 's/\\/\\\\/g' \
+    func_ar_run -LIST "$archive" | tr -d '\r' | sed -e 's/\\/\\\\/g' \
       | while read member
         do
-          $AR -NOLOGO -EXTRACT:"$member" "$archive" || exit $?
+          func_ar_run -EXTRACT:"$member" "$archive"
         done
   fi
 
@@ -262,14 +280,14 @@ elif test -n "$quick$replace"; then
   done
 
   if test -n "$orig_archive"; then
-    $AR -NOLOGO -OUT:"$archive" "$orig_archive" "$@" || exit $?
+    func_ar_run -OUT:${COLOR_MAJOR}"$archive"${COLOR_NONE} ${COLOR_MAJOR}"$orig_archive"${COLOR_NONE} "$@"
   else
-    $AR -NOLOGO -OUT:"$archive" "$@" || exit $?
+    func_ar_run -OUT:${COLOR_MAJOR}"$archive"${COLOR_NONE} "$@"
   fi
 
 elif test -n "$list"; then
   if test ! -f "$orig_archive"; then
     func_error "archive not found"
   fi
-  $AR -NOLOGO -LIST "$archive" || exit $?
+  func_ar_run -LIST ${COLOR_MAJOR}"$archive"${COLOR_NONE}
 fi
diff --git a/build-aux/compile b/build-aux/compile
index ba883b54aa..4e9157a535 100755
--- a/build-aux/compile
+++ b/build-aux/compile
@@ -1,5 +1,10 @@
 #! /bin/sh
+set -eo pipefail
+
 # Wrapper for compilers which do not understand '-c -o'.
+if [[ $GNU_BUILD_WRAPPER_DEBUG -eq 1 || -v COLOR_MAJOR ]]; then #if COLOR_MAJOR is set we want to include the wrapper so we can unset it otherwise would screw cmd output
+	. "$(dirname "$(readlink -f "$0")")/wrapper_helper.sh" "$@"
+fi
 
 scriptversion=2018-03-07.03; # UTC
 
@@ -49,6 +54,10 @@ func_file_conv ()
     / | /[!/]*) # absolute file, and not a UNC file
       if test -z "$file_conv"; then
 	# lazily determine how to convert abs files
+        OS_SYS=`uname -o`
+        if [[ "$OS_SYS" == "Msys" ]]; then
+            file_conv=cygwin
+        else
 	case `uname -s` in
 	  MINGW*)
 	    file_conv=mingw
@@ -61,6 +70,7 @@ func_file_conv ()
 	    ;;
 	esac
       fi
+      fi
       case $file_conv/,$2, in
 	*,$file_conv,*)
 	  ;;
@@ -91,37 +101,70 @@ func_cl_dashL ()
   linker_opts="$linker_opts -LIBPATH:$file"
 }
 
+function GetLibNameIfInDir() {
+	local lib=$1
+	local dir=$2
+	local dir_in_libpath=$3
+	local shared=$4
+	FULL_LIB_NAME=''
+
+	local TRY_NAMES=("${lib}.lib" "${lib}lib.lib" "lib${lib}.a" "lib${lib}.lib" "${lib}")
+	if $shared; then
+		TRY_NAMES+="${lib}.dll.lib"; #prefer dynamic last
+		#TRY_NAMES=("$lib.dll.lib" "${TRY_NAMES[@]}");
+	else
+		TRY_NAMES=("${lib}static.lib" "${lib}libstatic.lib" "${lib}-static.lib" "${lib}lib-static.lib" "${lib}_static.lib" "${lib}lib_static.lib" "${TRY_NAMES[@]}");
+	fi
+	for name in "${TRY_NAMES[@]}"; do
+		if test -f "${dir}/${name}"; then
+			if [ $dir_in_libpath -eq 1 ]; then
+				FULL_LIB_NAME="$name"
+			else
+				FULL_LIB_NAME="${dir}/${name}"
+			fi
+			return 0;
+		fi
+	done
+}
+
 # func_cl_dashl library
 # Do a library search-path lookup for cl
 func_cl_dashl ()
 {
-  lib=$1
-  found=no
-  save_IFS=$IFS
-  IFS=';'
-  for dir in $lib_path $LIB
-  do
-    IFS=$save_IFS
-    if $shared && test -f "$dir/$lib.dll.lib"; then
-      found=yes
-      lib=$dir/$lib.dll.lib
-      break
-    fi
-    if test -f "$dir/$lib.lib"; then
-      found=yes
-      lib=$dir/$lib.lib
-      break
-    fi
-    if test -f "$dir/lib$lib.a"; then
-      found=yes
-      lib=$dir/lib$lib.a
-      break
-    fi
-  done
+	lib=$1
+	found=no
+	save_IFS=$IFS
+	IFS=';'
+	FULL_LIB_NAME='' #incase no lib_paths we need to set this so the previous lib doesn't get used in the next check
+	for dir in $lib_path; do
+		IFS=$save_IFS
+		GetLibNameIfInDir "$lib" "$dir" 1 "$shared"
+		if [ ! -z "$FULL_LIB_NAME" ]; then
+			break
+		fi
+	done
+	if [ -z "$FULL_LIB_NAME" ]; then
+		IFS=';'
+		for dir in $LIB; do
+			IFS=$save_IFS
+			#seems like we should be able to leave off LIB dirs as cl.exe will incldue them anyway
+			GetLibNameIfInDir "$lib" "$dir" 1 "$shared"
+			if [ ! -z "$FULL_LIB_NAME" ]; then
+				break
+			fi
+		done
+	fi
+	if [ ! -z "$FULL_LIB_NAME" ]; then
+		found=yes
+		lib="$FULL_LIB_NAME"
+	fi
+
   IFS=$save_IFS
 
   if test "$found" != yes; then
-    lib=$lib.lib
+	if [[ "$lib" != *.lib ]]; then
+    	lib="$lib.lib"
+	fi	
   fi
 }
 
@@ -132,6 +175,7 @@ func_cl_wrapper ()
   # Assume a capable shell
   lib_path=
   shared=:
+  STD_DECLARE=""
   linker_opts=
   for arg
   do
@@ -145,16 +189,32 @@ func_cl_wrapper ()
 	  case $2 in
 	    *.o | *.[oO][bB][jJ])
 	      func_file_conv "$2"
-	      set x "$@" -Fo"$file"
+	      set x "$@" -Fo${COLOR_MAJOR}"$file"${COLOR_NONE}
 	      shift
 	      ;;
 	    *)
 	      func_file_conv "$2"
-	      set x "$@" -Fe"$file"
+	      set x "$@" -Fe${COLOR_MAJOR}"$file"${COLOR_NONE}
 	      shift
 	      ;;
 	  esac
 	  ;;
+	-include)
+		eat=1
+	  	set x "$@" -FI${COLOR_MINOR}"$2"${COLOR_NONE}
+	  	shift
+		;;
+	-x)
+	  eat=1
+	  LANG=$2
+	  if [[ "$LANG" == "c" ]]; then
+	  	LANG="-TC"
+	  elif [[ "$LANG" == "c++" ]]; then
+	  	LANG="-TP"
+	  fi
+	  set x "$@" "$LANG"
+	  shift
+	  ;;
 	-I)
 	  eat=1
 	  func_file_conv "$2" mingw
@@ -187,6 +247,21 @@ func_cl_wrapper ()
 	-static)
 	  shared=false
 	  ;;
+	-std=*)
+	  std="${1#-std=}"
+	  if [[ "$std" == "c99" ]]; then
+	  	std="c11"
+	  fi
+	  set x "$@" -std:"$std"
+	  if [[ "$std" == "c11" || "$std" == "c17" ]]; then #note that MSVC never supported dynamic arrays as it never supported c99 fully and c11 made them optional.  If your configure has a dynamic array test it will fail still.
+			STD_DECLARE="/Zc:__STDC__"
+	  fi
+	  shift
+	  ;;
+	-[Ff][pP][iI][cC] | -[Ff][Nn][Oo]-[Pp][Ii][Cc])
+	  ;;
+	-W[a-zA-Z][a-zA-Z-][a-zA-Z-][a-zA-Z-][a-zA-Z-]*) #word based warnings other than all are not supported by msvc
+	  ;;
 	-Wl,*)
 	  arg=${1#-Wl,}
 	  save_ifs="$IFS"; IFS=','
@@ -200,18 +275,25 @@ func_cl_wrapper ()
 	  eat=1
 	  linker_opts="$linker_opts $2"
 	  ;;
+	-version | --version)
+	  echo "MSVC CL" `cl 2>&1 1>/dev/null | grep -o -E "[0-9]+[.][0-9.]+"`
+	  exit
+	  ;;
 	-*)
 	  set x "$@" "$1"
 	  shift
 	  ;;
-	*.cc | *.CC | *.cxx | *.CXX | *.[cC]++)
+	*.cc | *.CC | *.cxx | *.CXX | *.[cC]++ | *.S)
 	  func_file_conv "$1"
-	  set x "$@" -Tp"$file"
+	  set x "$@" -Tp "$file"
 	  shift
 	  ;;
+	*.a)
+	  linker_opts="$linker_opts $1"
+	  ;;
 	*.c | *.cpp | *.CPP | *.lib | *.LIB | *.Lib | *.OBJ | *.obj | *.[oO])
 	  func_file_conv "$1" mingw
-	  set x "$@" "$file"
+	  set x "$@" "${COLOR_MINOR2}$file${COLOR_NONE}"
 	  shift
 	  ;;
 	*)
@@ -225,8 +307,12 @@ func_cl_wrapper ()
   if test -n "$linker_opts"; then
     linker_opts="-link$linker_opts"
   fi
-  exec "$@" $linker_opts
-  exit 1
+  if [[ $GNU_BUILD_WRAPPER_DEBUG -eq 1 ]]; then
+  	wrapper_exec "$@" $STD_DECLARE $linker_opts
+  else
+		exec "$@" $STD_DECLARE $linker_opts
+  fi
+
 }
 
 eat=
diff --git a/build-aux/ld-link b/build-aux/ld-link
new file mode 100644
index 0000000000..3b57d4e63b
--- /dev/null
+++ b/build-aux/ld-link
@@ -0,0 +1,45 @@
+#!/bin/bash
+set -eo pipefail
+
+# Wrapper for compilers which do not understand '-c -o'.
+if [[ $GNU_BUILD_WRAPPER_DEBUG -eq 1 || -v COLOR_MAJOR ]]; then #if COLOR_MAJOR is set we want to include the wrapper so we can unset it otherwise would screw cmd output
+	. "$(dirname "$(readlink -f "$0")")/wrapper_helper.sh" "$@"
+fi
+
+convert_from_msys_path () {
+	local WPATH=$1
+	if ! [ -z ${MSYS+x} ]; then
+		WPATH=`cygpath -m "$WPATH"`
+	fi
+	echo $WPATH
+}
+
+# with the other wrappers you can use cl or lib  as they genreally wont collide with msys but link.exe does 
+if [[ ! -v MS_LINK ]]; then
+	MS_LINK=`cygpath.exe -m "$VCToolsInstallDir"`
+	MS_LINK="${LINK_PATH}bin/HostX64/x64/link.exe"
+	if [[ ! -e "$MS_LINK" ]]; then
+		MS_LINK=`which link.exe -a | grep "MSVC" | head`
+		MS_LINK=$(convert_from_msys_path "$MS_LINK")
+	fi
+	if [[ ! -e "$MS_LINK" ]]; then
+		MS_LINK="link.exe"
+	fi
+fi
+
+for file in "$@"; do
+	if [[ $file = /[A-Za-z]/* ]]; then
+		file="${COLOR_MAJOR}\"$(convert_from_msys_path "$file")\"${COLOR_NONE}"
+	elif [[ -e "$file" ]]; then # we want to colorize above incase its the output file
+		file="${COLOR_MAJOR}\"$file\"${COLOR_NONE}"
+	fi
+	
+    set -- "$@" "$file"
+    shift
+done
+echo "HI MSLINK IS: $MS_LINK"
+if [[ $GNU_BUILD_WRAPPER_DEBUG -eq 1 ]]; then
+	wrapper_exec "$MS_LINK" -NOLOGO "$@"
+else
+	exec "$MS_LINK" -NOLOGO "$@"
+fi
diff --git a/build-aux/wrapper_helper.sh b/build-aux/wrapper_helper.sh
new file mode 100644
index 0000000000..22b2022db1
--- /dev/null
+++ b/build-aux/wrapper_helper.sh
@@ -0,0 +1,43 @@
+#!/bin/bash
+set -eo pipefail
+shopt -s inherit_errexit
+
+function clear_colors {
+	unset COLOR_MINOR COLOR_MINOR2 COLOR_MAJOR COLOR_NONE
+}
+
+CALLER_NAME=`basename ${0^^}`
+if [[ $GNU_BUILD_WRAPPER_DEBUG -eq 1 ]]; then
+	if [[ $GNU_BUILD_WRAPPER_COLOR -eq 	1 ]]; then
+		COLOR_MINOR="${COLOR_MINOR:-\e[2;33m}"
+		COLOR_MINOR2="${COLOR_MINOR2:-\e[2;36m}"
+		COLOR_MAJOR="${COLOR_MAJOR:-\e[1;32m}"
+		COLOR_NONE="${COLOR_NONE:-\e[0m}"
+	else
+		clear_colors;
+	fi
+	echo -e ${COLOR_MINOR}GNU ${CALLER_NAME} INPUT${COLOR_NONE}: "$@" 1>&2
+else
+	clear_colors;
+fi
+
+
+function wrapper_exec {	
+	shopt -s extglob
+	if [[ -v GNU_BUILD_CMD_FILE ]]; then
+		LAST_PWD_FILE="${GNU_BUILD_CMD_FILE}.tmpcurdir"
+		CUR_PWD=`pwd`
+		if [[ ! -e "$LAST_PWD_FILE" || "$CUR_PWD" != "`cat $LAST_PWD_FILE`" ]]; then
+			echo -e cd "$CUR_PWD" >> "$GNU_BUILD_CMD_FILE"
+			echo -n $CUR_PWD > "$LAST_PWD_FILE"
+		fi
+		echo -e "${@//'\e'\[*([0-9;])m/}" >> "$GNU_BUILD_CMD_FILE"
+	fi
+	if [[ $GNU_BUILD_WRAPPER_DEBUG -eq 1 ]]; then	
+		echo -e "${COLOR_MINOR}GNU ${CALLER_NAME} OUTPUT${COLOR_NONE}: " "$@" $STD_DECLARE $linker_opts 1>&2
+	fi
+	exec "${@//'\e'\[*([0-9;])m/}"  #strip ansi color strings out
+	shopt -u extglob
+
+}
+
diff --git a/doc/glibc-functions/endutent.texi b/doc/glibc-functions/endutent.texi
index b4f3dd397a..5e12e23c44 100644
--- a/doc/glibc-functions/endutent.texi
+++ b/doc/glibc-functions/endutent.texi
@@ -28,7 +28,4 @@ Portability problems not fixed by Gnulib:
 @item
 This function is missing on some platforms:
 macOS 11.1, FreeBSD 13.0, OpenBSD 6.7, Minix 3.1.8, mingw, MSVC 14.
-@item
-This function is not declared on some platforms:
-Android before ca.@: 2015.
 @end itemize
diff --git a/doc/glibc-functions/sysinfo.texi b/doc/glibc-functions/sysinfo.texi
index 16e12b12ac..3b1ade31a6 100644
--- a/doc/glibc-functions/sysinfo.texi
+++ b/doc/glibc-functions/sysinfo.texi
@@ -17,7 +17,4 @@ Portability problems not fixed by Gnulib:
 @item
 This function is missing on some platforms:
 macOS 11.1, FreeBSD 13.0, NetBSD 9.0, OpenBSD 6.7, Minix 3.1.8, AIX 5.1, Cygwin 1.7.9, mingw, MSVC 14.
-@item
-This function is not declared and thus not part of the Android API
-for Android API levels < 9.
 @end itemize
diff --git a/lib/boot-time-aux.h b/lib/boot-time-aux.h
index e59a0fd03c..348611fc85 100644
--- a/lib/boot-time-aux.h
+++ b/lib/boot-time-aux.h
@@ -65,7 +65,6 @@ get_linux_uptime (struct timespec *p_uptime)
     }
 # endif
 
-# if HAVE_DECL_SYSINFO /* not available in Android API < 9 */
   /* The sysinfo call returns the uptime with a resolution of 1 sec only.  */
   struct sysinfo info;
   if (sysinfo (&info) >= 0)
@@ -74,7 +73,6 @@ get_linux_uptime (struct timespec *p_uptime)
       p_uptime->tv_nsec = 0;
       return 0;
     }
-# endif
 
   return -1;
 }
diff --git a/lib/boot-time.c b/lib/boot-time.c
index fe5b5b88c8..d813bfa582 100644
--- a/lib/boot-time.c
+++ b/lib/boot-time.c
@@ -32,7 +32,7 @@
 # include <time.h>
 #endif
 
-#if HAVE_SYS_SYSCTL_H && !(defined __GLIBC__ && defined __linux__) && !defined __minix
+#if HAVE_SYS_SYSCTL_H && !defined __minix
 # if HAVE_SYS_PARAM_H
 #  include <sys/param.h>
 # endif
@@ -65,10 +65,8 @@
 # define UT_USER(UT) ((UT)->ut_user)
 #endif
 
-#if !HAVE_UTMPX_H && HAVE_UTMP_H && defined UTMP_NAME_FUNCTION
-# if !HAVE_DECL_ENDUTENT /* Android */
-void endutent (void);
-# endif
+#if !HAVE_UTMPX_H && HAVE_UTMP_H && defined UTMP_NAME_FUNCTION && !HAVE_DECL_GETUTENT
+struct utmp *getutent (void);
 #endif
 
 #if defined __linux__ || HAVE_UTMPX_H || HAVE_UTMP_H || defined __CYGWIN__ || defined _WIN32
diff --git a/lib/fts.c b/lib/fts.c
index faeb4b20a5..875fe05793 100644
--- a/lib/fts.c
+++ b/lib/fts.c
@@ -981,6 +981,7 @@ next:   tmp = p;
                         fts_load(sp, p);
                         if (! setup_dir(sp)) {
                                 free_dir(sp);
+                                __set_errno (ENOMEM);
                                 return (NULL);
                         }
                         goto check_for_dir;
@@ -1027,7 +1028,10 @@ check_for_dir:
                       sp->fts_dev = p->fts_statp->st_dev;
                     Dprintf (("  entering: %s\n", p->fts_path));
                     if (! enter_dir (sp, p))
-                      return NULL;
+                      {
+                        __set_errno (ENOMEM);
+                        return NULL;
+                      }
                   }
                 return p;
         }
@@ -1344,9 +1348,8 @@ fts_build (register FTS *sp, int type)
                   cur->fts_info = FTS_D;
                 else if (! enter_dir (sp, cur))
                   {
-                    int err = errno;
                     closedir_and_clear (cur->fts_dirp);
-                    __set_errno (err);
+                    __set_errno (ENOMEM);
                     return NULL;
                   }
               }
@@ -1934,7 +1937,6 @@ internal_function
 fts_lfree (register FTSENT *head)
 {
         register FTSENT *p;
-        int err = errno;
 
         /* Free a linked list of structures. */
         while ((p = head)) {
@@ -1943,8 +1945,6 @@ fts_lfree (register FTSENT *head)
                         closedir (p->fts_dirp);
                 free(p);
         }
-
-        __set_errno (err);
 }
 
 /*
diff --git a/lib/hash.c b/lib/hash.c
index a004885e28..918aa0d1c3 100644
--- a/lib/hash.c
+++ b/lib/hash.c
@@ -29,7 +29,6 @@
 #include "bitrotate.h"
 #include "xalloc-oversized.h"
 
-#include <errno.h>
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -500,17 +499,13 @@ compute_bucket_size (size_t candidate, const Hash_tuning *tuning)
     {
       float new_candidate = candidate / tuning->growth_threshold;
       if ((float) SIZE_MAX <= new_candidate)
-        goto nomem;
+        return 0;
       candidate = new_candidate;
     }
   candidate = next_prime (candidate);
   if (xalloc_oversized (candidate, sizeof (struct hash_entry *)))
-    goto nomem;
+    return 0;
   return candidate;
-
- nomem:
-  errno = ENOMEM;
-  return 0;
 }
 
 Hash_table *
@@ -539,7 +534,6 @@ hash_initialize (size_t candidate, const Hash_tuning *tuning,
          if the user provides invalid tuning options, we silently revert to
          using the defaults, and ignore further request to change the tuning
          options.  */
-      errno = EINVAL;
       goto fail;
     }
 
@@ -613,7 +607,6 @@ hash_free (Hash_table *table)
   struct hash_entry *bucket;
   struct hash_entry *cursor;
   struct hash_entry *next;
-  int err = errno;
 
   /* Call the user data_freer function.  */
   if (table->data_freer && table->n_entries)
@@ -656,8 +649,6 @@ hash_free (Hash_table *table)
   /* Free the remainder of the hash table structure.  */
   free (table->bucket);
   free (table);
-
-  errno = err;
 }
 
 /* Insertion and deletion.  */
@@ -771,8 +762,8 @@ hash_find_entry (Hash_table *table, const void *entry,
 /* Internal helper, to move entries from SRC to DST.  Both tables must
    share the same free entry list.  If SAFE, only move overflow
    entries, saving bucket heads for later, so that no allocations will
-   occur.  Return false (setting errno) if the free entry list is
-   exhausted and an allocation fails.  */
+   occur.  Return false if the free entry list is exhausted and an
+   allocation fails.  */
 
 static bool
 transfer_entries (Hash_table *dst, Hash_table *src, bool safe)
@@ -919,14 +910,12 @@ hash_rehash (Hash_table *table, size_t candidate)
      passes.  Two passes give worse cache performance and takes
      longer, but at this point, we're already out of memory, so slow
      and safe is better than failure.  */
-  int err = errno;
   table->free_entry_list = new_table->free_entry_list;
   if (! (transfer_entries (table, new_table, true)
          && transfer_entries (table, new_table, false)))
     abort ();
   /* table->n_entries already holds its value.  */
   free (new_table->bucket);
-  errno = err;
   return false;
 }
 
@@ -973,10 +962,7 @@ hash_insert_if_absent (Hash_table *table, void const *entry,
                 * tuning->growth_threshold));
 
           if ((float) SIZE_MAX <= candidate)
-            {
-              errno = ENOMEM;
-              return -1;
-            }
+            return -1;
 
           /* If the rehash fails, arrange to return NULL.  */
           if (!hash_rehash (table, candidate))
diff --git a/lib/hash.h b/lib/hash.h
index af07242ff4..54d3f22f69 100644
--- a/lib/hash.h
+++ b/lib/hash.h
@@ -148,7 +148,7 @@ typedef void (*Hash_data_freer) (void *entry);
 /* Reclaim all storage associated with a hash table.  If a data_freer
    function has been supplied by the user when the hash table was created,
    this function applies it to the data of each entry before freeing that
-   entry.  This function preserves errno, like 'free'.  */
+   entry.  */
 extern void hash_free (Hash_table *table);
 
 /* Allocate and return a new hash table, or NULL upon failure.  The initial
@@ -183,9 +183,7 @@ extern void hash_free (Hash_table *table);
    You should specify this function only if you want these functions to free
    all of your 'data' data.  This is typically the case when your data is
    simply an auxiliary struct that you have malloc'd to aggregate several
-   values.
-
-   Set errno on failure; otherwise errno is unspecified.  */
+   values.  */
 _GL_ATTRIBUTE_NODISCARD
 extern Hash_table *hash_initialize (size_t candidate,
                                     const Hash_tuning *tuning,
@@ -194,7 +192,7 @@ extern Hash_table *hash_initialize (size_t candidate,
                                     Hash_data_freer data_freer)
   _GL_ATTRIBUTE_MALLOC _GL_ATTRIBUTE_DEALLOC (hash_free, 1);
 
-/* Like hash_initialize, but invokes xalloc_die instead of returning NULL.  */
+/* Same as hash_initialize, but invokes xalloc_die on memory exhaustion.  */
 /* This function is defined by module 'xhash'.  */
 _GL_ATTRIBUTE_NODISCARD
 extern Hash_table *hash_xinitialize (size_t candidate,
@@ -220,26 +218,25 @@ extern void hash_clear (Hash_table *table);
    the table may receive at least CANDIDATE different user entries, including
    those already in the table, before any other growth of the hash table size
    occurs.  If TUNING->IS_N_BUCKETS is true, then CANDIDATE specifies the
-   exact number of buckets desired.  Return true iff the rehash succeeded,
-   false (setting errno) otherwise.  */
+   exact number of buckets desired.  Return true iff the rehash succeeded.  */
 _GL_ATTRIBUTE_NODISCARD
 extern bool hash_rehash (Hash_table *table, size_t candidate);
 
 /* If ENTRY matches an entry already in the hash table, return the pointer
    to the entry from the table.  Otherwise, insert ENTRY and return ENTRY.
-   Return NULL (setting errno) if the storage required for insertion
-   cannot be allocated.  This implementation does not support
-   duplicate entries or insertion of NULL.  */
+   Return NULL if the storage required for insertion cannot be allocated.
+   This implementation does not support duplicate entries or insertion of
+   NULL.  */
 _GL_ATTRIBUTE_NODISCARD
 extern void *hash_insert (Hash_table *table, const void *entry);
 
-/* Same as hash_insert, but invokes xalloc_die instead of returning NULL.  */
+/* Same as hash_insert, but invokes xalloc_die on memory exhaustion.  */
 /* This function is defined by module 'xhash'.  */
 extern void *hash_xinsert (Hash_table *table, const void *entry);
 
 /* Insert ENTRY into hash TABLE if there is not already a matching entry.
 
-   Return -1 (setting errno) upon memory allocation failure.
+   Return -1 upon memory allocation failure.
    Return 1 if insertion succeeded.
    Return 0 if there is already a matching entry in the table,
    and in that case, if MATCHED_ENT is non-NULL, set *MATCHED_ENT
diff --git a/lib/nproc.c b/lib/nproc.c
index e3de1873a9..2740c458c1 100644
--- a/lib/nproc.c
+++ b/lib/nproc.c
@@ -46,7 +46,7 @@
 # include <sys/param.h>
 #endif
 
-#if HAVE_SYS_SYSCTL_H && !(defined __GLIBC__ && defined __linux__)
+#if HAVE_SYS_SYSCTL_H && ! defined __GLIBC__
 # include <sys/sysctl.h>
 #endif
 
@@ -306,7 +306,7 @@ num_processors_ignoring_omp (enum nproc_query query)
   /* Finally, as fallback, use the APIs that don't distinguish between
      NPROC_CURRENT and NPROC_ALL.  */
 
-#if HAVE_SYSCTL && !(defined __GLIBC__ && defined __linux__) && defined HW_NCPU
+#if HAVE_SYSCTL && ! defined __GLIBC__ && defined HW_NCPU
   { /* This works on macOS, FreeBSD, NetBSD, OpenBSD.
        macOS 10.14 does not allow mib to be const.  */
     int nprocs;
diff --git a/lib/physmem.c b/lib/physmem.c
index 398f99b727..f450587141 100644
--- a/lib/physmem.c
+++ b/lib/physmem.c
@@ -50,7 +50,7 @@
 # include <sys/param.h>
 #endif
 
-#if HAVE_SYS_SYSCTL_H && !(defined __GLIBC__ && defined __linux__)
+#if HAVE_SYS_SYSCTL_H && ! defined __GLIBC__
 # include <sys/sysctl.h>
 #endif
 
@@ -96,7 +96,7 @@ double
 physmem_total (void)
 {
 #if defined _SC_PHYS_PAGES && defined _SC_PAGESIZE
-  { /* This works on linux-gnu, kfreebsd-gnu, solaris2, and cygwin.  */
+  { /* This works on linux-gnu, solaris2 and cygwin.  */
     double pages = sysconf (_SC_PHYS_PAGES);
     double pagesize = sysconf (_SC_PAGESIZE);
     if (0 <= pages && 0 <= pagesize)
@@ -153,8 +153,8 @@ physmem_total (void)
   }
 #endif
 
-#if HAVE_SYSCTL && !(defined __GLIBC__ && defined __linux__) && defined HW_PHYSMEM
-  { /* This works on *bsd, kfreebsd-gnu, and darwin.  */
+#if HAVE_SYSCTL && ! defined __GLIBC__ && defined HW_PHYSMEM
+  { /* This works on *bsd and darwin.  */
     unsigned int physmem;
     size_t len = sizeof physmem;
     static int mib[2] = { CTL_HW, HW_PHYSMEM };
@@ -208,7 +208,7 @@ double
 physmem_available (void)
 {
 #if defined _SC_AVPHYS_PAGES && defined _SC_PAGESIZE
-  { /* This works on linux-gnu, kfreebsd-gnu, solaris2, and cygwin.  */
+  { /* This works on linux-gnu, solaris2 and cygwin.  */
     double pages = sysconf (_SC_AVPHYS_PAGES);
     double pagesize = sysconf (_SC_PAGESIZE);
     if (0 <= pages && 0 <= pagesize)
@@ -267,8 +267,8 @@ physmem_available (void)
   }
 #endif
 
-#if HAVE_SYSCTL && !(defined __GLIBC__ && defined __linux__) && defined HW_USERMEM
-  { /* This works on *bsd, kfreebsd-gnu, and darwin.  */
+#if HAVE_SYSCTL && ! defined __GLIBC__ && defined HW_USERMEM
+  { /* This works on *bsd and darwin.  */
     unsigned int usermem;
     size_t len = sizeof usermem;
     static int mib[2] = { CTL_HW, HW_USERMEM };
diff --git a/lib/readutmp.c b/lib/readutmp.c
index 7967db60a8..ef9f0aff43 100644
--- a/lib/readutmp.c
+++ b/lib/readutmp.c
@@ -40,7 +40,7 @@
 # include <systemd/sd-login.h>
 #endif
 
-#if HAVE_SYS_SYSCTL_H && !(defined __GLIBC__ && defined __linux__) && !defined __minix
+#if HAVE_SYS_SYSCTL_H && !defined __minix
 # if HAVE_SYS_PARAM_H
 #  include <sys/param.h>
 # endif
@@ -314,11 +314,9 @@ have_boot_time (struct utmp_alloc a)
   return false;
 }
 
-#if !HAVE_UTMPX_H && HAVE_UTMP_H && defined UTMP_NAME_FUNCTION
-# if !HAVE_DECL_ENDUTENT /* Android */
-void endutent (void);
+# if !HAVE_UTMPX_H && HAVE_UTMP_H && defined UTMP_NAME_FUNCTION && !HAVE_DECL_GETUTENT
+struct utmp *getutent (void);
 # endif
-#endif
 
 static int
 read_utmp_from_file (char const *file, idx_t *n_entries, STRUCT_UTMP **utmp_buf,
diff --git a/lib/readutmp.h b/lib/readutmp.h
index 3ddecf3727..1cf588d265 100644
--- a/lib/readutmp.h
+++ b/lib/readutmp.h
@@ -136,18 +136,14 @@ enum { UT_HOST_SIZE = -1 };
 /* This is a near-copy of glibc's struct utmpx, which stops working
    after the year 2038.  Unlike the glibc version, struct utmpx32
    describes the file format even if time_t is 64 bits.  */
-#define _GL_UT_USER_SIZE  sizeof (((struct utmpx *) 0)->ut_user)
-#define _GL_UT_ID_SIZE    sizeof (((struct utmpx *) 0)->ut_id)
-#define _GL_UT_LINE_SIZE  sizeof (((struct utmpx *) 0)->ut_line)
-#define _GL_UT_HOST_SIZE  sizeof (((struct utmpx *) 0)->ut_host)
 struct utmpx32
 {
   short int ut_type;               /* Type of login.  */
   pid_t ut_pid;                    /* Process ID of login process.  */
-  char ut_line[_GL_UT_LINE_SIZE];  /* Devicename.  */
-  char ut_id[_GL_UT_ID_SIZE];      /* Inittab ID.  */
-  char ut_user[_GL_UT_USER_SIZE];  /* Username.  */
-  char ut_host[_GL_UT_HOST_SIZE];  /* Hostname for remote login. */
+  char ut_line[__UT_LINESIZE];     /* Devicename.  */
+  char ut_id[4];                   /* Inittab ID.  */
+  char ut_user[__UT_USERSIZE];     /* Username.  */
+  char ut_host[__UT_HOSTSIZE];     /* Hostname for remote login. */
   struct __exit_status ut_exit;    /* Exit status of a process marked
                                       as DEAD_PROCESS.  */
   /* The fields ut_session and ut_tv must be the same size when compiled
@@ -253,13 +249,6 @@ struct utmpx32
 # define WTMP_FILE "/etc/wtmp"
 #endif
 
-/* In early versions of Android, <utmp.h> did not define BOOT_TIME, only
-   USER_PROCESS.  We need to use the value that is defined in newer versions
-   of Android.  */
-#if defined __ANDROID__ && !defined BOOT_TIME
-# define BOOT_TIME 2
-#endif
-
 /* Some platforms, such as OpenBSD, don't have an ut_type field and don't have
    the BOOT_TIME and USER_PROCESS macros.  But we want to support them in
    'struct gl_utmp'.  */
diff --git a/m4/readutmp.m4 b/m4/readutmp.m4
index 0a47f4bb77..fff8d4eb7b 100644
--- a/m4/readutmp.m4
+++ b/m4/readutmp.m4
@@ -1,4 +1,4 @@
-# readutmp.m4 serial 30
+# readutmp.m4 serial 28
 dnl Copyright (C) 2002-2023 Free Software Foundation, Inc.
 dnl This file is free software; the Free Software Foundation
 dnl gives unlimited permission to copy and/or distribute it,
@@ -55,7 +55,7 @@ AC_DEFUN_ONCE([gl_PREREQ_READUTMP_H],
   if test $ac_cv_header_utmp_h = yes || test $ac_cv_header_utmpx_h = yes; then
     dnl Prerequisites of lib/readutmp.h and lib/readutmp.c.
     AC_CHECK_FUNCS_ONCE([utmpname utmpxname])
-    AC_CHECK_DECLS([endutent],,,[[
+    AC_CHECK_DECLS([getutent],,,[[
 /* <sys/types.h> is a prerequisite of <utmp.h> on FreeBSD 8.0, OpenBSD 4.6.  */
 #include <sys/types.h>
 #ifdef HAVE_UTMP_H
@@ -103,10 +103,6 @@ AC_INCLUDES_DEFAULT
     AC_CHECK_MEMBERS([struct utmp.ut_exit.e_termination],,,[$utmp_includes])
   fi
 
-  AC_CHECK_DECLS([sysinfo],,,[[
-    #include <sys/sysinfo.h>
-    ]])
-
   AC_CHECK_HEADERS_ONCE([sys/param.h])
   dnl <sys/sysctl.h> requires <sys/param.h> on OpenBSD 4.0.
   AC_CHECK_HEADERS([sys/sysctl.h],,,
diff --git a/m4/stdint.m4 b/m4/stdint.m4
index b9f764d4c1..d6961b0993 100644
--- a/m4/stdint.m4
+++ b/m4/stdint.m4
@@ -1,4 +1,4 @@
-# stdint.m4 serial 62
+# stdint.m4 serial 61
 dnl Copyright (C) 2001-2023 Free Software Foundation, Inc.
 dnl This file is free software; the Free Software Foundation
 dnl gives unlimited permission to copy and/or distribute it,
@@ -150,10 +150,7 @@ intmax_t i = INTMAX_MAX;
 uintmax_t j = UINTMAX_MAX;
 
 /* Check that SIZE_MAX has the correct type, if possible.  */
-/* ISO C 11 mandates _Generic, but GCC versions < 4.9 lack it.  */
-#if 201112 <= __STDC_VERSION__ \
-    && (!defined __GNUC__ || 4 < __GNUC__ + (9 <= __GNUC_MINOR__) \
-        || defined __clang__)
+#if 201112 <= __STDC_VERSION__
 int k = _Generic (SIZE_MAX, size_t: 0);
 #elif (2 <= __GNUC__ || 4 <= __clang_major__ || defined __IBM__TYPEOF__ \
        || (0x5110 <= __SUNPRO_C && !__STDC__))
diff --git a/modules/boot-time-tests b/modules/boot-time-tests
index 9436f0e8b5..bb56ee36e5 100644
--- a/modules/boot-time-tests
+++ b/modules/boot-time-tests
@@ -1,6 +1,5 @@
 Files:
 tests/test-boot-time.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/dprintf-gnu-tests b/modules/dprintf-gnu-tests
index 176d64c10c..117e615649 100644
--- a/modules/dprintf-gnu-tests
+++ b/modules/dprintf-gnu-tests
@@ -2,7 +2,6 @@ Files:
 tests/test-dprintf-gnu.sh
 tests/test-dprintf-gnu.c
 tests/test-printf-gnu.output
-tests/macros.h
 
 Depends-on:
 dprintf-posix-tests
diff --git a/modules/fclose-tests b/modules/fclose-tests
index 6e8ebe8343..0f36e0d50e 100644
--- a/modules/fclose-tests
+++ b/modules/fclose-tests
@@ -1,6 +1,5 @@
 Files:
 tests/test-fclose.c
-tests/macros.h
 
 Depends-on:
 fdopen
diff --git a/modules/fts b/modules/fts
index b754e0f572..fe56bae6e0 100644
--- a/modules/fts
+++ b/modules/fts
@@ -11,7 +11,6 @@ Depends-on:
 alignasof
 attribute
 c99
-calloc-posix
 closedir
 cycle-check
 d-ino
@@ -20,19 +19,16 @@ fchdir
 fcntl
 fcntl-h
 flexmember
-free-posix
 fstat
 hash
 i-ring
 lstat
-malloc-posix
 memmove
 open
 openat-h
 opendir
 opendirat
 readdir
-realloc-posix
 stdbool
 stddef
 stdint
diff --git a/modules/getcwd-tests b/modules/getcwd-tests
index 4a8aee2540..c842c77115 100644
--- a/modules/getcwd-tests
+++ b/modules/getcwd-tests
@@ -2,7 +2,6 @@ Files:
 tests/test-getcwd.sh
 tests/test-getcwd.c
 tests/qemu.h
-tests/macros.h
 
 Depends-on:
 errno
diff --git a/modules/hash b/modules/hash
index e837317d9f..42502e749e 100644
--- a/modules/hash
+++ b/modules/hash
@@ -7,9 +7,6 @@ lib/hash.h
 
 Depends-on:
 bitrotate
-calloc-posix
-free-posix
-malloc-posix
 stdbool
 stdint
 xalloc-oversized
diff --git a/modules/lock-tests b/modules/lock-tests
index 502c36882f..31d627da2a 100644
--- a/modules/lock-tests
+++ b/modules/lock-tests
@@ -3,7 +3,6 @@ tests/test-rwlock1.c
 tests/test-lock.c
 tests/test-once.c
 tests/atomic-int-gnulib.h
-tests/macros.h
 m4/semaphore.m4
 
 Depends-on:
diff --git a/modules/nullptr-c++-tests b/modules/nullptr-c++-tests
index 391881974d..ae60f4b360 100644
--- a/modules/nullptr-c++-tests
+++ b/modules/nullptr-c++-tests
@@ -1,6 +1,5 @@
 Files:
 tests/test-nullptr-c++.cc
-tests/macros.h
 
 Status:
 c++-test
diff --git a/modules/physmem-tests b/modules/physmem-tests
deleted file mode 100644
index c37c41cd7e..0000000000
--- a/modules/physmem-tests
+++ /dev/null
@@ -1,11 +0,0 @@
-Files:
-tests/test-physmem.c
-tests/macros.h
-
-Depends-on:
-
-configure.ac:
-
-Makefile.am:
-TESTS += test-physmem
-check_PROGRAMS += test-physmem
diff --git a/modules/posix_spawn-tests b/modules/posix_spawn-tests
index ef56cf6023..7d316c96f6 100644
--- a/modules/posix_spawn-tests
+++ b/modules/posix_spawn-tests
@@ -8,7 +8,6 @@ tests/executable-script
 tests/executable-script.sh
 tests/executable-shell-script
 tests/signature.h
-tests/macros.h
 
 Depends-on:
 posix_spawn_file_actions_init
diff --git a/modules/posix_spawnp-tests b/modules/posix_spawnp-tests
index 16369ff52f..4ac569e318 100644
--- a/modules/posix_spawnp-tests
+++ b/modules/posix_spawnp-tests
@@ -8,7 +8,6 @@ tests/executable-script
 tests/executable-script.sh
 tests/executable-shell-script
 tests/signature.h
-tests/macros.h
 
 Depends-on:
 posix_spawn_file_actions_init
diff --git a/modules/readutmp-tests b/modules/readutmp-tests
index c21fe9c4be..5543b30fac 100644
--- a/modules/readutmp-tests
+++ b/modules/readutmp-tests
@@ -1,6 +1,5 @@
 Files:
 tests/test-readutmp.c
-tests/macros.h
 
 Depends-on:
 array-mergesort
diff --git a/modules/sh-quote-tests b/modules/sh-quote-tests
index 1d72a1899d..b2eccc58c7 100644
--- a/modules/sh-quote-tests
+++ b/modules/sh-quote-tests
@@ -1,6 +1,5 @@
 Files:
 tests/test-sh-quote.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/u16-grapheme-breaks-tests b/modules/unigbrk/u16-grapheme-breaks-tests
index 4051d3cdea..9f7f2dffef 100644
--- a/modules/unigbrk/u16-grapheme-breaks-tests
+++ b/modules/unigbrk/u16-grapheme-breaks-tests
@@ -1,6 +1,5 @@
 Files:
 tests/unigbrk/test-u16-grapheme-breaks.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/u16-grapheme-next-tests b/modules/unigbrk/u16-grapheme-next-tests
index 724f37d858..a947afe3f3 100644
--- a/modules/unigbrk/u16-grapheme-next-tests
+++ b/modules/unigbrk/u16-grapheme-next-tests
@@ -1,6 +1,5 @@
 Files:
 tests/unigbrk/test-u16-grapheme-next.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/u16-grapheme-prev-tests b/modules/unigbrk/u16-grapheme-prev-tests
index 1f126032de..be62d2468a 100644
--- a/modules/unigbrk/u16-grapheme-prev-tests
+++ b/modules/unigbrk/u16-grapheme-prev-tests
@@ -1,6 +1,5 @@
 Files:
 tests/unigbrk/test-u16-grapheme-prev.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/u32-grapheme-breaks-tests b/modules/unigbrk/u32-grapheme-breaks-tests
index 50119c8b5d..4561344744 100644
--- a/modules/unigbrk/u32-grapheme-breaks-tests
+++ b/modules/unigbrk/u32-grapheme-breaks-tests
@@ -1,6 +1,5 @@
 Files:
 tests/unigbrk/test-u32-grapheme-breaks.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/u32-grapheme-next-tests b/modules/unigbrk/u32-grapheme-next-tests
index a5e5d8b887..4ed8b1f94e 100644
--- a/modules/unigbrk/u32-grapheme-next-tests
+++ b/modules/unigbrk/u32-grapheme-next-tests
@@ -1,6 +1,5 @@
 Files:
 tests/unigbrk/test-u32-grapheme-next.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/u32-grapheme-prev-tests b/modules/unigbrk/u32-grapheme-prev-tests
index ac76585b0b..0900722791 100644
--- a/modules/unigbrk/u32-grapheme-prev-tests
+++ b/modules/unigbrk/u32-grapheme-prev-tests
@@ -1,6 +1,5 @@
 Files:
 tests/unigbrk/test-u32-grapheme-prev.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/u8-grapheme-breaks-tests b/modules/unigbrk/u8-grapheme-breaks-tests
index a9e6991ce3..7d1f6d98d5 100644
--- a/modules/unigbrk/u8-grapheme-breaks-tests
+++ b/modules/unigbrk/u8-grapheme-breaks-tests
@@ -1,6 +1,5 @@
 Files:
 tests/unigbrk/test-u8-grapheme-breaks.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/u8-grapheme-next-tests b/modules/unigbrk/u8-grapheme-next-tests
index 2c676c61f1..d30709ea63 100644
--- a/modules/unigbrk/u8-grapheme-next-tests
+++ b/modules/unigbrk/u8-grapheme-next-tests
@@ -1,6 +1,5 @@
 Files:
 tests/unigbrk/test-u8-grapheme-next.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/u8-grapheme-prev-tests b/modules/unigbrk/u8-grapheme-prev-tests
index 7c830946cf..137e8464f7 100644
--- a/modules/unigbrk/u8-grapheme-prev-tests
+++ b/modules/unigbrk/u8-grapheme-prev-tests
@@ -1,6 +1,5 @@
 Files:
 tests/unigbrk/test-u8-grapheme-prev.c
-tests/macros.h
 
 Depends-on:
 
diff --git a/modules/unigbrk/uc-grapheme-breaks-tests b/modules/unigbrk/uc-grapheme-breaks-tests
index 2b0493930e..3c211ff5cb 100644
--- a/modules/unigbrk/uc-grapheme-breaks-tests
+++ b/modules/unigbrk/uc-grapheme-breaks-tests
@@ -2,7 +2,6 @@ Files:
 tests/unigbrk/test-uc-grapheme-breaks.c
 tests/unigbrk/test-uc-grapheme-breaks.sh
 tests/unigbrk/GraphemeBreakTest.txt
-tests/macros.h
 
 Depends-on:
 
diff --git a/tests/test-physmem.c b/tests/test-physmem.c
deleted file mode 100644
index 232cc86d12..0000000000
--- a/tests/test-physmem.c
+++ /dev/null
@@ -1,38 +0,0 @@
-/* Test of getting the amount of total/available physical memory.
-   Copyright (C) 2023 Free Software Foundation, Inc.
-
-   This file is free software: you can redistribute it and/or modify
-   it under the terms of the GNU General Public License as published
-   by the Free Software Foundation, either version 3 of the License,
-   or (at your option) any later version.
-
-   This file is distributed in the hope that it will be useful,
-   but WITHOUT ANY WARRANTY; without even the implied warranty of
-   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-   GNU General Public License for more details.
-
-   You should have received a copy of the GNU General Public License
-   along with this program.  If not, see <https://www.gnu.org/licenses/>.  */
-
-/* Written by Bruno Haible <bruno@clisp.org>, 2023.  */
-
-#include <config.h>
-
-#include "physmem.h"
-
-#include <stdio.h>
-
-#include "macros.h"
-
-int
-main (int argc, char *argv[])
-{
-  printf ("Total memory:     %12.f B = %6.f MiB\n",
-          physmem_total (), physmem_total () / (1024.0 * 1024.0));
-  printf ("Available memory: %12.f B = %6.f MiB\n",
-          physmem_available (), physmem_available () / (1024.0 * 1024.0));
-  ASSERT (physmem_total () >= physmem_available ());
-  ASSERT (physmem_available () >= 4 * 1024 * 1024);
-
-  return 0;
-}
